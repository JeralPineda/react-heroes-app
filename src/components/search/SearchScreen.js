import React, { useMemo } from 'react';
import queryString from 'query-string';

import { useLocation } from 'react-router';
import { HeroCard } from '../heroes/HeroCard';
import { useForm } from '../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';

import './search.css';

export const SearchScreen = ({ history }) => {
   // obtener la location desde los componentes(url)
   const location = useLocation();
   const { q = '' } = queryString.parse(location.search);

   const [formValues, handleInputChange] = useForm({
      searchHero: q,
   });

   const { searchHero } = formValues;

   const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

   //    Filtro de búsqueda
   const handleSearch = (e) => {
      e.preventDefault();

      //   query de busqueda
      history.push(`?q=${searchHero}`);
   };

   return (
      <div>
         <h1>Search Screen</h1>
         <hr />

         <div className='row'>
            <div className='col-5'>
               <h4>Search Form</h4>
               <hr />

               <form onSubmit={handleSearch}>
                  <input className='form-control defaultInput' type='text' placeholder='Busca tu héroe' autoComplete='off' name='searchHero' value={searchHero} onChange={handleInputChange} />
               </form>

               <button type='submit' className='btn m-1 btn-block btn-outline-secondary' onClick={handleSearch}>
                  Buscar...
               </button>
            </div>

            <div className='col-7'>
               <h4>Results</h4>
               <hr />
               {/* no mostrar caja de cards si esta vacio el input */}

               {q === '' && <div className='alert alert-info'>Busca un heroe</div>}

               {q !== '' && heroesFiltered.length === 0 && <div className='alert alert-danger'>No hay un héroe con {q}</div>}

               {heroesFiltered.map((hero) => (
                  <HeroCard key={hero.id} {...hero} />
               ))}
            </div>
         </div>
      </div>
   );
};
