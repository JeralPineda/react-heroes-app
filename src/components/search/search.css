.alert-info {
   background-color: rgb(197, 199, 206) !important;
   color: #222;
}
