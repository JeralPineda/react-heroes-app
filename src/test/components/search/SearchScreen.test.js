import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('Pruebas en <SearchScreen />', () => {
   test('debe de mostrarse correctamente con valores por defecto', () => {
      const wrapper = mount(
         <MemoryRouter initialEntries={['/search']}>
            <Route path='/search' component={SearchScreen} />
         </MemoryRouter>
      );

      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find('.alert-info').text().trim()).toBe('Busca un heroe');
   });

   test('debe de mostrar a Flash y el input con el valor del queryString', () => {
      const wrapper = mount(
         <MemoryRouter initialEntries={['/search?q=flash']}>
            <Route path='/search' component={SearchScreen} />
         </MemoryRouter>
      );

      expect(wrapper.find('input').prop('value')).toBe('flash');
      expect(wrapper).toMatchSnapshot();
   });

   test('debe de mostrar un error si no se muestra el Hero', () => {
      const wrapper = mount(
         <MemoryRouter initialEntries={['/search?q=flash123']}>
            <Route path='/search' component={SearchScreen} />
         </MemoryRouter>
      );

      expect(wrapper.find('.alert-danger').text().trim()).toBe(`No hay un héroe con flash123`);
      expect(wrapper).toMatchSnapshot();
   });

   test('debe de llamar el push del history', () => {
      const history = {
         push: jest.fn(),
      };

      const wrapper = mount(
         <MemoryRouter initialEntries={['/search?q=flash123']}>
            <Route path='/search' component={() => <SearchScreen history={history} />} />
         </MemoryRouter>
      );

      //   Simulación cambio en la caja de texto
      wrapper.find('input').simulate('change', {
         target: {
            name: 'searchHero',
            value: 'flash',
         },
      });

      //   Simulación del formulario
      wrapper.find('form').prop('onSubmit')({
         preventDefault() {},
      });

      expect(history.push).toHaveBeenCalledWith(`?q=flash`);
   });
});
