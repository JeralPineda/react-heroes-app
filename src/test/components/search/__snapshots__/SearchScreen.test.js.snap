// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> debe de mostrar a Flash y el input con el valor del queryString 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control defaultInput"
          name="searchHero"
          onChange={[Function]}
          placeholder="Busca tu héroe"
          type="text"
          value="flash"
        />
      </form>
      <button
        className="btn m-1 btn-block btn-outline-secondary"
        onClick={[Function]}
        type="submit"
      >
        Buscar...
      </button>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card card-color mb-1 ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Flash"
              className="card-img"
              src="./heroes/dc-flash.jpg"
            />
          </div>
          <div
            className="cold-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Flash
              </h5>
              <p
                className="card-text"
              >
                Jay Garrick
              </p>
               && 
              <p
                className="card-text"
              >
                Jay Garrick, Barry Allen, Wally West, Bart Allen
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Flash Comics #1
                </small>
              </p>
              <a
                className="link"
                href="/hero/dc-flash"
                onClick={[Function]}
              >
                Mas ..
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> debe de mostrar un error si no se muestra el Hero 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control defaultInput"
          name="searchHero"
          onChange={[Function]}
          placeholder="Busca tu héroe"
          type="text"
          value="flash123"
        />
      </form>
      <button
        className="btn m-1 btn-block btn-outline-secondary"
        onClick={[Function]}
        type="submit"
      >
        Buscar...
      </button>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        No hay un héroe con 
        flash123
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control defaultInput"
          name="searchHero"
          onChange={[Function]}
          placeholder="Busca tu héroe"
          type="text"
          value=""
        />
      </form>
      <button
        className="btn m-1 btn-block btn-outline-secondary"
        onClick={[Function]}
        type="submit"
      >
        Buscar...
      </button>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Busca un heroe
      </div>
    </div>
  </div>
</div>
`;
